src/drivers/amd/renoir/acp_dmic_dma.c

```diff
@@ -0,0 +1,378 @@
+#include <platform/fw_scratch_mem.h>
+#include <platform/chip_registers.h>
+
+static uint32_t dmic_rngbuff_size;
+/* 109c7aba-a7ba-43c3-b9-42-59-e2-0a-66-11-be */
+DECLARE_SOF_UUID("acp_dmic_dma", acp_dmic_dma_uuid, 0x109c7aba, 0xa7ba, 0x43c3,
+		0xb9, 0x42, 0x59, 0xe2, 0x0a, 0x66, 0x11, 0xbe);
+DECLARE_TR_CTX(acp_dmic_dma_tr, SOF_UUID(acp_dmic_dma_uuid), LOG_LEVEL_INFO);
+
+/* allocate next free DMA channel */
+static struct dma_chan_data *acp_dmic_dma_channel_get(struct dma *dma,
+						   unsigned int req_chan)
+{
+	uint32_t flags;
+	struct dma_chan_data *channel;
+
+	spin_lock_irq(&dma->lock, flags);
+	if (req_chan >= dma->plat_data.channels) {
+		spin_unlock_irq(&dma->lock, flags);
+		tr_err(&acp_dmic_dma_tr, "ACP_DMIC: Channel %d out of range",
+				req_chan);
+		return NULL;
+	}
+	channel = &dma->chan[req_chan];
+	if (channel->status != COMP_STATE_INIT) {
+		spin_unlock_irq(&dma->lock, flags);
+		tr_err(&acp_dmic_dma_tr, "ACP_DMIC: Cannot reuse channel %d",
+				req_chan);
+		return NULL;
+	}
+	atomic_add(&dma->num_channels_busy, 1);
+	channel->status = COMP_STATE_READY;
+	spin_unlock_irq(&dma->lock, flags);
+	return channel;
+}
+
+static void acp_dmic_dma_channel_put(struct dma_chan_data *channel)
+{
+	uint32_t flags;
+
+	notifier_unregister_all(NULL, channel);
+	spin_lock_irq(&channel->dma->lock, flags);
+	channel->status = COMP_STATE_INIT;
+	atomic_sub(&channel->dma->num_channels_busy, 1);
+	spin_unlock_irq(&channel->dma->lock, flags);
+}
+
+static int acp_dmic_dma_start(struct dma_chan_data *channel)
+{
+	acp_wov_pdm_no_of_channels_t pdm_channels;
+	acp_wov_pdm_decimation_factor_t deci_fctr;
+	acp_wov_clk_ctrl_t clk_ctrl;
+	acp_wov_misc_ctrl_t wov_misc_ctrl;
+	acp_wov_pdm_dma_enable_t  pdm_dma_enable;
+
+	channel->status = COMP_STATE_ACTIVE;
+	if (channel->direction == DMA_DIR_DEV_TO_MEM) {
+		/* Channel for DMIC */
+		pdm_channels.bits.pdm_no_of_channels = 0;
+		io_reg_write(PU_REGISTER_BASE + ACP_WOV_PDM_NO_OF_CHANNELS,
+							pdm_channels.u32all);
+		/* Decimation Factor */
+		deci_fctr.u32all = 2;
+		io_reg_write(PU_REGISTER_BASE + ACP_WOV_PDM_DECIMATION_FACTOR,
+							deci_fctr.u32all);
+		/* DMIC Clock */
+		clk_ctrl.bits.brm_clk_ctrl = 7;
+		io_reg_write(PU_REGISTER_BASE + ACP_WOV_CLK_CTRL,
+						clk_ctrl.u32all);
+		/* PDM Control */
+		wov_misc_ctrl = (acp_wov_misc_ctrl_t)io_reg_read(PU_REGISTER_BASE + ACP_WOV_MISC_CTRL);
+		wov_misc_ctrl.u32all |= 0x10;
+		io_reg_write(PU_REGISTER_BASE + ACP_WOV_MISC_CTRL,
+						wov_misc_ctrl.u32all);
+		/* PDM Enable */
+		io_reg_write(PU_REGISTER_BASE + ACP_WOV_PDM_ENABLE, 1);
+		/* PDM DMA Enable */
+		io_reg_write(PU_REGISTER_BASE + ACP_WOV_PDM_DMA_ENABLE, 1);
+		/* Check the PDM DMA Enable status bit */
+		pdm_dma_enable = (acp_wov_pdm_dma_enable_t)
+			io_reg_read(PU_REGISTER_BASE + ACP_WOV_PDM_DMA_ENABLE);
+		while (!(uint32_t)pdm_dma_enable.bits.pdm_dma_en_status) {
+			pdm_dma_enable = (acp_wov_pdm_dma_enable_t)
+			io_reg_read(PU_REGISTER_BASE + ACP_WOV_PDM_DMA_ENABLE);
```

Comment:
safety counter?

Reply:
Addressed the Review Comment. Added safety timer counter to avoide infinite while loop.